File: 프로그래머스/lv2/최원준/쿼드압축 후 개수 세기.py
'''
1. 아이디어 :
    dfs로 풀 수 있다
2. 시간복잡도 :
    O( n^2 log n )
3. 자료구조 :
    -
'''

def solution(arr):

    def recur(x, y, n):
        base = arr[x][y]
        for row in range(x, x+n):
            for col in range(y, y+n):
                if arr[row][col] != base:
                    n = n // 2
                    recur(x, y, n) # topleft
                    recur(x+n, y, n) # botleft
                    recur(x, y+n, n) # topright
                    recur(x+n, y+n, n) #botright
                    return
        ans[base] += 1

    ans = [0,0]
    recur(0, 0, len(arr))
    return ans

File: 프로그래머스/lv2/지준혁/쿼드 압축 후 개수 세기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

bool is_compress(int n, int r, int c, vector<vector<int>>& arr, int& zero, int& one) {
    int st = arr[r][c];
    for (int i = r; i < n + r; ++i) {
        for (int j = c; j < n + c; ++j) {
            if (arr[i][j] != st) return false;
        }
    }
    st == 0 ? zero++ : one++;
    return true;
}

void recur(int n, int r, int c, vector<vector<int>>& arr, int& zero, int& one) {
    
    if (n == 0 || is_compress(n, r, c, arr, zero, one)) {
        
        return; 
    }
    int half = n / 2;
    recur(half, r, c, arr, zero, one);
    recur(half, r, c + half, arr, zero, one);
    recur(half, r + half, c, arr, zero, one); 
    recur(half, r + half, c + half, arr, zero, one);
}
    
vector<int> solution(vector<vector<int>> arr) {
    vector<int> ans = {0 , 0};
    recur(arr.size(), 0, 0, arr, ans[0], ans[1]);
    return ans;
}
